designs/2019-move-to-async-api/README.md

```diff
@@ -0,0 +1,531 @@
+- [static compareResultsByFilePath()](#-new-methods) (new)
+
+Initially the `ESLint` class will be a wrapper around `CLIEngine`, modifying return types. Later it can take on a more independent shape as `CLIEngine` gets more deprecated.
+
+#### ● Constructor
+
+The constructor has mostly the same options as `CLIEngine`, but with some differences:
+
+- It throws fatal errors if the options contain unknown properties or an option is invalid type ([eslint/eslint#10272](https://github.com/eslint/eslint/issues/10272)).
+- It disallows the deprecated `cacheFile` option.
+- It has a new `pluginImplementations` option as the successor of `addPlugin()` method. This is an object that keys are plugin IDs and each value is the plugin object. See also "[The other methods](#-the-other-methods)" section.
+
+<details>
+<summary>A rough sketch of the constructor.</summary>
+
+```js
+class ESLint {
+  constructor({
+    allowInlineConfig = true,
+    baseConfig = null,
+    cache = false,
+    cacheLocation = ".eslintcache",
+    configFile = null,
+    cwd = process.cwd(),
+    envs = [],
+    extensions = [".js"],
+    fix = false,
+    fixTypes = ["problem", "suggestion", "layout"],
+    globInputPaths = true,
+    globals = [],
+    ignore = true,
+    ignorePath = null,
+    ignorePattern = [],
+    parser = "espree",
+    parserOptions = null,
+    pluginImplementations = null,
+    plugins = [],
+    reportUnusedDisableDirectives = false,
+    resolvePluginsRelativeTo = cwd,
+    rulePaths = [],
+    rules = null,
+    useEslintrc = true,
+    ...unknownOptions
+  } = {}) {
+    // Throws on unknown options
+    if (Object.keys(unknownOptions).length >= 1) {
+      //...
+    }
+    // Throws on the invalid value of options
+    if (typeof allowInlineConfig !== "boolean") {
+      // ...
+    }
+    if (typeof baseConfig !== "object") {
+      // ...
+    }
+    // and other options...
+
+    // Initialize CLIEngine because this is a tiny wrapper.
+    const engine = (this._cliEngine = new CLIEngine({
+      allowInlineConfig,
+      baseConfig,
+      cache,
+      cacheLocation,
+      configFile,
+      cwd,
+      envs,
+      extensions,
+      fix,
+      fixTypes,
+      globInputPaths,
+      globals,
+      ignore,
+      ignorePath,
+      ignorePattern,
+      parser,
+      parserOptions,
+      plugins,
+      reportUnusedDisableDirectives,
+      resolvePluginsRelativeTo,
+      rulePaths,
+      rules,
+      useEslintrc,
+    }))
+
+    // Apply `pluginImplementations` option.
+    if (pluginImplementations) {
+      for (const [id, definition] of Object.entries(pluginImplementations)) {
+        engine.addPlugin(id, definition)
+      }
+    }
+  }
+}
+```
+
+</details>
+
+#### ● The `executeOnFiles()` method
+
+This method returns an object that implements [AsyncIterable] and [AsyncIterator], as similar to async generators. Therefore we can use the returned object with `for-await-of` statement.
```

Comment:
See end note about async iteration approach.

Reply:
> I question whether progressive formatters even make sense for ESLint. As you mentioned in this proposal, we cannot guarantee the order in which file linting will be executed, so do users really want individual updates when files can come out in any order? (Early on this was also the case and then we changed the results to sorted alphabetically due to user complaints.)

I believe so. Progressive formatters are a big win.

Because the current ESLint looks freeze silently while linting. For example, in my environment, when I lint our codebase (eslint/eslint), the `eslint` command spends 20 seconds with no output. I cannot distinguish if the `eslint` command is working fine or frozen for some reason. With larger codebase, it may be multiple minutes.

Instead, our default formatter can print progress: ![image](https://github.com/eslint/rfcs/raw/formatter-v2/designs/2019-formatter-v2/images/stylish.gif)

Therefore, users can be calm by seeing the `eslint` is working clearly.

Of course, most formatters may sort the result. However, I believe it's good if it's open to implementing streaming for custom formatter.

> Plus, most formatters have a specific format that needs a defined start and end (like JUnit XML and JSON) where it would never make sense to output anything progressively in case there's a crash.

This is not a problem. In most cases, the start and end are the same stuff regardless of the lint results. Formatting is easy:

```js
let first = true
yield "["
for await (const result of results) {
    if (!first) {
        first = false
        yield ","
    }
    yield JSON.stringify(result)
}
yield "]"
```

> But if we really wanted to provide insights into what ESLint is doing, we can always turn CLIEngine into an event emitter and emit "lintstart", "lintfinish", etc. events that give people insights into what's happening without the need for async iteration. That would be enough for API consumers to output progress bars if they so choose without forcing everyone through a more complicated process.

I don't think that making CLIEngine to an event emitter for this purpose makes sense. Because it's unique for each the `executeOnFiles()` method call rather than the `CLIEngine` instance. API users have to distinguish where the events came. Plus, we have to design how we notify those events to formatters in our own way. I believe that it makes sense if we use the standard way in the language rather than the specific way per tool.

Reply:
What if we just created a new API that allows users to opt in to event emitting?

Something like:

```js
const cliEngine = new CLIEngine();
const lintRunner = cliEngine.createNewLintRunner(/* text/file options */);

lintRunner.on("fileStart", fileInfo => {});

lintRunner.run();
```

Reply:
My point is not that progress indicators wouldn't be useful (your gif is nice!), it's that I don't see the value of embedding that functionality in formatters (making them more complicated) when that could just be handled in the CLI by us all the time. 

For example, that progress indicator you showed could just be built into ESLint by default. It could run when ESLint starts, disappear when linting is complete, and then run formatters as we do today.

I just don't think there's enough evidence that this functionality needs to be built into formatters.

>  I believe that it makes sense if we use the standard way in the language rather than the specific way per tool.

Again, I feel like async iteration is too confusing and error-prone for API consumers. Events are a much easier to understand model for what is going on here. You do make a good point about knowing where the events came from, so we could return an event emitter from `executeOnFiles()`:

```js
const session = engine.executeOnFiles(filesToLint);
session.on("start", ({totalFileCount}) => {});
session.on("filestart", ({filePath}) => {});
session.on("filecomplete", ({filePath}) => {});
session.on("complete", ({fileCount}) => {});
session.on("error", (error) => {});
```

This would allow everything to happen asynchronously and the promise would resolve to the `complete` event and reject to the `error` event.

Reply:
@platinumazure same wavelength :)

Reply:
> For example, that progress indicator you showed could just be built into ESLint by default. It could run when ESLint starts, disappear when linting is complete, and then run formatters as we do today.

I think it makes sense if the formatter handles all outputs because people need to have varied progress displays. For example, people may like progress bars in common. But CI's websites don't support CSI sequence (i.e. progress bar) so people may like another kind of progress display. In another case, people may like to show the location of errors on demand to reduce waiting time. And another case, people don't want to show progress because a program parses the output.

If the formatters can output progress state, they can make favorite progress by custom formatters. Otherwise, ESLint has to support the option that selects the kind of progress along with custom formatters.

> I feel like async iteration is too confusing and error-prone for API consumers. Events are a much easier to understand model for what is going on here.

It's hard for me to agree that async iterator is error-prone and event emitter is not. We need to set up on the object relatively more than async iterator. I predict that API users will request to implement AsyncIterable interface and Thennable interface to the session object to handle `filestart` event by `for-await-of` loops and to handle `complete` and `error` events by `await` expressions. (Similarly, the Node.js streams have been implemented AsyncIterable interface.)

Instead, would we start with a minimal spec?

The `executeOnFiles()` returns the object that implements only AsyncIterable and Thennable. So users can use it with either `async` expressions and `for-await-of` loop.

```js
// With await
const results = await engine.executeOnFiles(filesToLint);

// With for-await-of
for await (const result of engine.executeOnFiles(filesToLint)) {
}

// A bonus, streams
const { Readable } = require("stream")
const resultStream = Readable.from(engine.executeOnFiles(filesToLint))
resultStream
    .on("data", result => {})
    .on("end", () => {})
    .on("error", error => {})
```

We can add events later if needed.

Reply:
> I think it makes sense if the formatter handles all outputs because people need to have varied progress displays

I think it's a mistake to bundle that requirement with this proposal, as I think it's clouding the issue of the correct interface to use to replace `CLIEngine`. We can easily provide a default progress indicator that doesn't require any changes to how we do formatters. If, at some later point, we start getting requests to customize the progress indicator, we can always design something for that. 

>  I predict that API users will request to implement AsyncIterable interface and Thennable interface to the session object to handle the filecomplete events by for-await-of loops and to handle the complete and error events by await expressions. (Similarly, the Node.js streams have been implemented AsyncIterable interface.)

I just don't think either async iteration or streams makes sense for this API. I'm having a hard time imagining why something like Standard or Visual Studio Code would want anything less than all of the lint results. The only thing you've suggested as a use case for partial results is a progress indicator, which again, I just don't think is important enough to design the API around.

Reply:
> I just don't think either async iteration or streams makes sense for this API.

I believe that to support `for-await-of` makes sense for this API.

```js
// Get results at once.
const results = await engine.lintFiles(patterns);
// Get results step by step.
for await (const result of engine.lintFiles(patterns)) {
}
```

This symmetry is clearly. The result is a list and we can iterate the list. If we stop the iteration by control statements (`break`, `return`, or `throw`), it just cancels the rest of linting. Much intuitive.

> I'm having a hard time imagining why something like Standard or Visual Studio Code would want anything less than all of the lint results. 

Editor integrations lint each file on shown or typed, so I guess that they don't use async iteration.

CLI wrappers may want to use async iteration because they sometimes have linted large codebase. To show the message and location on demand is worthful in order to know the error location as soon (rather than after waiting a few minutes).
This is the same for us.

> The only thing you've suggested as a use case for partial results is a progress indicator, which again, I just don't think is important enough to design the API around.

Why do you think it's not important?

> I think it's a mistake to bundle that requirement with this proposal, as I think it's clouding the issue of the correct interface to use to replace CLIEngine. We can easily provide a default progress indicator that doesn't require any changes to how we do formatters. If, at some later point, we start getting requests to customize the progress indicator, we can always design something for that.

It's bundled in this proposal because it affects the design of public API:

1. `engine.lintFiles(patterns)` should return results rather than output to stdout, even if about progress state.
1. So the output of progress should happen between the `engine.lintFiles(patterns)` and the formatter's process, or in the formatter's process.
1. `engine.getFormatter()` API has to change by whether the formatter's process outputs about progress or not. We cannot do it without more breaking changes.

Currently, it's hard to update our custom formatter spec because of the breaking changes of `getFormatter()`. If the `engine.getFormatter()` API returns the adapter of formatters, probably we can update our custom formatter spec without breaking changes of the `engine.getFormatter()` API in the future. Related to this, #45 depends on this proposal.

Reply:
I think that sorting lint results is the responsibility of formatters rather than the `lintFiles()` method. The `lintFiles()` method should focus on linting then up to formatters how it outputs the results, includes the order of outputs.

Reply:
OK, I will remove about progress printing from this proposal.